### PR TITLE
[DataGrid] Implement internationalization

### LINKED ISSUE #196: [DataGrid] Implement i18n
### Benchmark

- https://demos.telerik.com/kendo-ui/grid/localization
- https://material-table.com/#/docs/features/localization
- https://www.telerik.com/kendo-react-ui/components/grid/globalization/
- https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/localization/
- https://datatables.net/manual/i18n
- http://tabulator.info/docs/4.5/localize
- https://www.jqwidgets.com/react/react-grid/#https://www.jqwidgets.com/react/react-grid/react-grid-localization.htm
- https://ag-grid.com/javascript-grid-localisation/

### PR DESCRIPTION
**Pull request by DanailH**
One iteration on #196 

### Overview
I'm this is the first part of the "Implement Localization" ticket. For the second part there is going to be a follow up ticket. This feature is available on both the `DataGrid` and `XGrid` components.

### What does this PR solves?
```
- We add a localeText prop to the DataGrid and XGrid components. This is an object that can be globally overridden through the theme with theme.components.MuiDataGrid.defaultProps.localeText.
- The prop performs a deep merge with the default translations.
- The default translations are in the en_US locale.
- We use the context to propagate the prop down to all the components of the grid. This propagation can be implemented with a custom context or the apiRef, to be determined on what we need to optimize. No matter the solution we pick, it needs to propagate everywhere in the data grid. I memo shouldn't prevent the value to be updated.
- We document it.
```

I've added a `useI18n` hook that has only one exported method for now that is the `getText`. It is used to get the text value that is assigned to a key in the 'localText' object.

Later on we can extend the hook with additional functionality needed for part 2.

Documentation, tests and full text replacement is missing from this Draft. Once we have the implementation and the **naming** 😅   nailed down I'll add in the missing things and convert the Draft into a normal PR.

### Example usage:

```jsx
<DataGrid localeText={{ densityText: 'size' }} />
```

### ToDo:
- [x] Add one smoke test
- [x] Update docs https://material-ui.com/components/data-grid/localization/
- [x] Update API pages https://material-ui.com/api/data-grid/, https://material-ui.com/api/x-grid/
- [x] Use `getText` to display the text in the grid (on one component initially, a follow up PR will be created for the rest of the components)

### CODE REVIEW

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx [hunk 1]
```diff
@@ -25,17 +25,17 @@ export function DensitySelector() {
   const DensityOptions: Array<DensityOption> = [
     {
       icon: <DensityCompactIcon />,
-      label: 'Compact',
+      label: apiRef?.current.getText('densityOptionCompact') as string,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:28]: I don't think that we can accept a component to **ever** render without its label, so `!` instead of `?` so it crashes if it ever happens. If it happens, I strongly think that we need to change the architecture, e.g. move to using the context or else.

```suggestion
      label: apiRef!.current.getText('densityOptionCompact') as string,
```

File: packages/storybook/src/stories/grid-i18n.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  title: 'X-Grid Tests/i18n',
+  component: XGrid,
+  parameters: {
+    options: { selectedPanel: 'storybook/storysource/panel' },
+    docs: {
+      page: null,
+    },
+  },
+};
+export const CustomLocaleText = () => {
```
Comment (oliviertassinari) [inline packages/storybook/src/stories/grid-i18n.stories.tsx:16]: I don't think that we need to store for this. A demo in the documentation to iterate on it, and a smoke end-to-end test should be enough. I think that it's important we work on the documentation by default, only move to storybook when it's no longer enough. So far, stories are meant for edge cases we can't cover in the documentation. We have talked about extending for

- rendering end-to-end tests
- rendering custom visual regression tests.


```suggestion
```

**Review by oliviertassinari — COMMENTED** [review]
Great idea to only translate a few keys!

File: packages/storybook/src/stories/grid-i18n.stories.tsx (hunk 2 above)
Comment (DanailH) [reply packages/storybook/src/stories/grid-i18n.stories.tsx:16]: Ok, noted. I usually add the story first because I use the storybook as a dev env. I'll create a demo in the docs. I see that there is already a page for it so I will update it 😉 [(DataGrid Localization)](https://material-ui.com/components/data-grid/localization/#data-grid-localization)

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 1 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:28]: I blame autocomplete.

File: packages/storybook/src/stories/grid-i18n.stories.tsx (hunk 2 above)
Comment (dtassone) [reply packages/storybook/src/stories/grid-i18n.stories.tsx]: > we work on the documentation first

Well, I don't think we should do the mentioned workflow. 
1. We add issues in the docs due to an extra plugin, so the environment is not properly isolated. And it's babel and not TS...
2. We render multiple demos on one page so, the logs are going to mix between grids, then breakpoints hit... 
3. The storybook seems faster to reload, and it does not need to reload the full page. Also is there type checking in docs?
In demos, you write in TS and JS at the same time, will you transpile manually at each change?

IMHO, you cannot build high-quality enterprise products properly in the middle of a next js doc website. 
The purpose of the storybook is to facilitate the development of components, providing tools, isolation, and all the required environment to successfully develop components.

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx [hunk 3]
```diff
@@ -25,17 +25,17 @@ export function DensitySelector() {
   const DensityOptions: Array<DensityOption> = [
     {
       icon: <DensityCompactIcon />,
-      label: 'Compact',
+      label: apiRef!.current.getText('densityOptionCompact') as string,
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:28]: Could we remove as string?

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 4]
```diff
@@ -357,6 +358,10 @@ export interface GridOptions {
    * Set the density of the grid.
    */
   density: Density;
+  /**
+   * Set the locale text of the grid.
+   */
+  localeText: object;
```
Comment (dtassone) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:364]: How about we type a bit better?
`{[key: string]: string}` or `Record<string, string>`

File: packages/grid/_modules_/grid/constants/i18nConstants.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export const DEFAULT_LOCALE_TEXT = {
```
Comment (dtassone) [inline packages/grid/_modules_/grid/constants/i18nConstants.ts:1]: How about we store that in a separate JSON file? So we can a file per culture?

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 3 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:28]: This is just a type fix as the Label receives either a string or undefined.

File: packages/storybook/src/stories/grid-i18n.stories.tsx (hunk 2 above)
Comment (oliviertassinari) [reply packages/storybook/src/stories/grid-i18n.stories.tsx]: There is probably a balance to find between the two workflows. I see use cases for both. My concern was, at its core, that **when it's possible**, it's much better to start with the documentation. Given that we have a standalone documentation page for the localization. I thought that it was viable. At the end of the day, pick what works best for you. However, I think that we should be aware that by not working on the documentation first, we also make it more likely that the DX in the documentation won't be as great as it could have been. We have seen this problem happen a. couple of times so far with the data grid, where we introduced regressions. We should be careful to not make the docs a second-class citizen. The documentation is very important (based on the survey we did). So I think that it should be handled at the same time with any new feature as we do it for tests.

1. I don't follow everything.
  - What extra plugin? We have been working like this for years and at a very large scale in the main repo. I can safely say that it's working great. The root of the issue we got with the babel plugin doing the memoization was with the logic of the data grid, the docs, only surface it. We will have to fix this root issue anyway for the component composable API, then we can add the babel plugin back.
  - The isolated environement can be harmful too. It making us blind to how the components are used in real life. We have seen this happen multiple times on the data grid. e.g. wrong resize position, leaking keyboard isolation.
  - Why does it matter that it's Babel, not TS? The plan is to use the core bundling strategy with the enterprise components, they diverge from the convention. TS is great for static analysis, there is no point in using it for bundling, as far as I understand it.
2. I agree that working in isolation also has its use cases. How often do we work with the logs enabled? 
3.
 - Agree, the full reload of the documentation page sucks. I have no idea why it happens on material-ui-x. It would be great to fix it, to get the same DX as in the main repository.
 - For the JavaScript demos, the current workflow is to only work on the TypeScript ones. Once you are done (e.g. after a first review), you run the script that outputs the JavaScript ones. There is also a watch command to do such but I have never used it.

File: packages/grid/_modules_/grid/constants/i18nConstants.ts (hunk 5 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/constants/i18nConstants.ts:1]: Sure, where would you want to store that json? We can keep it in the `constants/` folder for now?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/constants/i18nConstants.ts:1]: @dtassone You should find more details on the linked issue #196. The globalization topic was split into two phases 1. internationalization (to empower localization) 2. localization. For now, we are working on the internationalization infrastructure (1.).

Your comment is on 2. the plan was to do it later. We wanted to open a new issue for it once #196 is closed with this pull request. We can continue the conversation. For 2, we should be able to empower the same approach to what we do with the other components for consistency.

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 3 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:28]: Mind that sometimes, we might want to allow React nodes, not just string

Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx]: Done

File: packages/grid/_modules_/grid/models/gridOptions.tsx (hunk 4 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: done

File: packages/grid/_modules_/grid/constants/i18nConstants.ts (hunk 5 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/constants/i18nConstants.ts:1]: > Sure, where would you want to store that json? We can keep it in the constants/ folder for now?

Please don't, not yet.

File: packages/grid/_modules_/grid/models/gridOptions.tsx (hunk 4 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: We will also need to support react node

Comment (DanailH) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: I've created a custom type to cover the 3 different cases -> `string` | `React.ReactNode` | `Function`. The only inconvenience with this is that depending on the types  we may need to cast it depending on the place where it is used, like in here `label: apiRef!.current.getText('densityOptionCompact') as string`. I really don't want to change the typings of existing interfaces and also that is more a corner case. In the majority of the time plain text is used directly, so it wouldn't be a problem.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: @DanailH What do you think about listing all the keys?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: I would expect most of the keys to be react node, sometime we might only support a string when it needs to be applied on an HTML attribute. We might also need a function, from time to time. You can check the type of the translated strings in the existing components of the main repository to get an idea on the frequency of each case.

Comment (DanailH) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: We can definitely list all the keys but this list will grow over time. The approach of generalising looked simpler at the time. I'll check the typings in the main repo as well.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: I think that from a DX perspective, listing all the keys will be better. The overhead for us seems limited.

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx [hunk 6]
```diff
@@ -25,17 +25,17 @@ export function DensitySelector() {
...
       value: DensityTypes.Compact,
     },
     {
       icon: <DensityStandardIcon />,
-      label: 'Standard',
+      label: apiRef!.current.getText('densityOptionStandard') as string,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:33]: Do we need a getter? Why not access the prop directly?

File: packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+export const useI18n = (apiRef: ApiRef): void => {
+  const logger = useLogger('useI18n');
+  const { localeText } = useGridSelector(apiRef, optionsSelector);
+
+  const getText = React.useCallback(
+    (key: string): LocaleTextValue => {
+      logger.debug(`Get grid text with key ${key}`);
+
+      return localeText[key] || '';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx:17]:
```suggestion
      if (process.env.NODE_ENV !== 'production' && localeText.getOwnProperty(key)) {
        throw new Error(`Wrong key used for internal purposes ${key}`);
      }

      return localeText[key];
```

File: packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import { I18nApi, LocaleTextValue } from '../../../models/api/i18nApi';
+import { useApiMethod } from '../../root/useApiMethod';
+import { useGridSelector } from '../core/useGridSelector';
+import { optionsSelector } from '../../utils/useOptionsProp';
+
+export const useI18n = (apiRef: ApiRef): void => {
+  const logger = useLogger('useI18n');
+  const { localeText } = useGridSelector(apiRef, optionsSelector);
+
+  const getText = React.useCallback(
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx:13]:
```suggestion
  const getText = React.useCallback(
```

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx:13]: Nothing is change here

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 6 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:33]: I would say yes - the reason is to be aligned with the general approach of implementing features in the grid.

File: packages/grid/_modules_/grid/constants/i18nConstants.ts (hunk 5 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/constants/i18nConstants.ts:1]: I'm resolving - the outcome is that we will keep it in a js file because we may need to use either functions or nodes for some of the cases

File: packages/grid/_modules_/grid/models/gridOptions.tsx (hunk 4 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/models/gridOptions.tsx]: I think we cleared out the typings - I've updated the code with the new typings. I'll resolve this and if need to discuss it again we can open a new discussion thread.

**Comment by DanailH** [comment]
I've fixed the outstanding issues and comments, added a test case and updated the docs. 

Note: I've only converted the text in the Density Selector component for now to make the PR more focused on the functionality. Once this is merged I'll open a follow up PR that handles the rest of the text used in the grid.

File: docs/pages/api-docs/data-grid.md [hunk 9]
```diff
@@ -59,6 +59,7 @@ import { DataGrid } from '@material-ui/data-grid';
 | <span class="prop-name">onPageSizeChange</span> | <span class="prop-type">(param: PageChangeParams) => void</span> |   | Callback fired when the page size has changed. |
 | <span class="prop-name">onError</span> | <span class="prop-type">(args: any) => void</span> |   | Callback fired when an exception is thrown in the grid, or when the `showError` API method is called. |
 | <span class="prop-name">icons</span> | <span class="prop-type">IconsOptions</span> |   | Set of icons used in the grid. |
+| <span class="prop-name">localeText</span> | <span class="prop-type">LocaleText</span> |   | Set of text labels used in the grid. |
```
Comment (oliviertassinari) [inline docs/pages/api-docs/data-grid.md:62]: sort asc, but you will need to rebase first as there is a conflict

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 10]
```diff
@@ -87,6 +88,18 @@ export enum DensityTypes {
   Comfortable = 'comfortable',
 }
 
+/**
+ * Sets the types of the texts in the grid.
+ */
+export interface LocaleText {
+  // Density selector toolbar button text
+  densityText?: string | React.ReactNode;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:96]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2338e34f760b10c7599d22a58ce41295a9d7c286/types/react/index.d.ts#L232-L237

```suggestion
  densityText?: React.ReactNode;
```

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 11]
```diff
@@ -360,4 +360,23 @@ describe('<DataGrid /> - Layout & Warnings', () => {
...
+
+  describe('i18n', () => {
+    before(function beforeHook() {
+      if (/jsdom/.test(window.navigator.userAgent)) {
+        // Need layouting
+        this.skip();
+      }
+    });
+
+    it('should replace the Density Selector button label text to "Size"', () => {
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:372]:
```suggestion
    it('should replace the density selector button label text to "Size"', () => {
```

File: packages/grid/_modules_/grid/models/api/i18nApi.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * The i18n API interface that is available in the grid [[apiRef]].
+ */
+export interface I18nApi {
+  /**
+   * Get grid text.
+   * @param key
+   * @returns LocaleTextValue
+   */
+  getText: (key: string) => LocaleTextValue;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/api/i18nApi.ts:12]: I think that this type is wrong, it should retain the type of the key in LocaleText.

File: packages/grid/_modules_/grid/constants/i18nConstants.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+  densityText: 'Density',
+  densityLabelText: 'Density',
+  densityOptionTextCompact: 'Compact',
+  densityOptionTextStandard: 'Standard',
+  densityOptionTextComfortable: 'Comfortable',
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/constants/i18nConstants.ts:6]: Do we need to mention "text" in the key? I think that the name of the prop is enough. The benchmark links can give some inspiration on the naming convention. For instance, some use a nested structure, ag-grid is flat on the other hand.

Maybe we could also prefix with the context "toolbar" and be shorter on the place where it's used, focused more on the actual meaning:

```suggestion
  toolbarDensity: 'Density',
  toolbarDensityLabel: 'Density',
  toolbarDensityCompact: 'Compact',
  toolbarDensityStandard: 'Standard',
  toolbarDensityComfortable: 'Comfortable',
```

File: docs/src/pages/components/data-grid/localization/localization.md [hunk 14]
```diff
@@ -7,15 +7,15 @@ components: DataGrid, XGrid
...
-
-> ⚠️ This feature isn't implemented yet. It's coming.
->
-> 👍 Upvote [issue #196](https://github.com/mui-org/material-ui-x/issues/196) if you want to see it land faster.
+## Locale text
 
 The grid will support localization with a single prop that includes all the translated messages.
 You will be able to follow [this guide](/guides/localization/#locale-text) to use them.
 
+You can use the `localeText` prop to pass in your own text and translations. The default translations are for en_EN.
```
Comment (mbrookes) [inline docs/src/pages/components/data-grid/localization/localization.md:15]:
```suggestion
You can use the `localeText` prop to pass in your own text and translations. The default translations are for en_US.
```

**Review by mbrookes — DISMISSED** [review]

File: packages/grid/_modules_/grid/models/api/i18nApi.ts (hunk 12 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/models/api/i18nApi.ts:12]: This only defines what can be returned. I can remove the return type and put it as any but other than that I don't think I can make it dependent on the type of the object key. I don't have access to `key` when determining the return type

File: packages/grid/_modules_/grid/constants/i18nConstants.ts (hunk 13 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/constants/i18nConstants.ts:6]: I think I like this more, It does bring more context regarding where that locale text is being used

File: packages/grid/_modules_/grid/models/api/i18nApi.ts (hunk 12 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/api/i18nApi.ts]: I have pushed a change that restricts the types to be closer to the underlying implementation.

**Review by oliviertassinari — APPROVED** [review]

File: packages/grid/_modules_/grid/models/api/i18nApi.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import { DEFAULT_LOCALE_TEXT } from '../../constants/i18nConstants';
+
+export type LocaleTextValue = string | React.ReactNode | Function;
+
+export type Translations = typeof DEFAULT_LOCALE_TEXT;
```
Comment (dtassone) [inline packages/grid/_modules_/grid/models/api/i18nApi.ts:5]: I think it would be cleaner to use an interface instead of the default object.

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx [hunk 16]
```diff
@@ -23,17 +23,17 @@ export function DensitySelector() {
...
     {
       icon: <DensityStandardIcon />,
-      label: 'Standard',
+      label: apiRef!.current.getText('toolbarDensityStandard'),
       value: DensityTypes.Standard,
     },
     {
       icon: <DensityComfortableIcon />,
-      label: 'Comfortable',
+      label: apiRef!.current.getText('toolbarDensityComfortable'),
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:36]: Not sure about the api method name. It seems a bit too generic?
How about we rename it to `getLocaleText`? Also we could rename the hook to `useLocale ` if it handles node and strings. We could also create 2 methods, one for node and one for text ie getLocaleText, `getLocaleNode` so it can be used only in specific edge cases. And we can avoid errors where the node is returned instead of a plain text.
We can also consider using `i18n` instead of `locale`

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 17]
```diff
@@ -87,6 +88,18 @@ export enum DensityTypes {
   Comfortable = 'comfortable',
 }
 
+/**
+ * Sets the types of the texts in the grid.
+ */
+export interface LocaleText {
+  // Density selector toolbar button text
```
Comment (dtassone) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:95]: That's the interface I mean

Comment (dtassone) [reply packages/grid/_modules_/grid/models/gridOptions.tsx:95]: It should be a model

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 16 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:36]: > How about we rename it to getLocaleText?

Sounds great. 

> We could also create 2 methods,

Why would it be better with 2 methods? What about when we need to ask for a function, do we add a third method? The returned type is already correct so I don't see  how it would help avoid issues more than we currently have.

File: packages/grid/_modules_/grid/models/api/i18nApi.ts (hunk 15 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/models/api/i18nApi.ts:5]: If it is taken from the interface because the properties are defined as optional the return type results in `something | undefined` and then we need to add `as something` when the method is called in certain places. The thing is that the entries needs to be optional so that users can pass only parts of the translations if they want.

File: packages/grid/_modules_/grid/models/gridOptions.tsx (hunk 17 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/models/gridOptions.tsx:95]: Changing it to a type will still result in the problem mentioned above, because entries will still be optional.

File: packages/grid/_modules_/grid/models/api/i18nApi.ts (hunk 15 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/api/i18nApi.ts:5]: Oops, yes, my bad, I should have used LocaleText, I was trying to make the POC work 😁

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 16 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:36]: I can rename it to `getI18n` but I would prefer to have only one method since actually both of them will do the same thing. The only difference will be that we how going to have 2 API methods to maintain and to choose which one to use based on the useless. It's much simpler to have 1 that only selects an entry from an object and returns it.

For `i18n` instead of 'locale' are you talking about the prop name? Like `i18nText` or just `i18n` instead of `localeText`?

File: packages/grid/_modules_/grid/models/api/i18nApi.ts (hunk 15 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/models/api/i18nApi.ts:5]: > If it is taken from the interface because the properties are defined as optional the return type results in `something | undefined` and then we need to add `as something` when the method is called in certain places. The thing is that the entries needs to be optional so that users can pass only parts of the translations if they want.

Use a `Partial<LocaleText>` in the prop, like for `GridOptions`

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 16 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:36]: To avoid, having 2 method then the interface should have a determined type for each key

```
interface LocaleText {
 labelKey: string;
 ChildrenValue: String | React.Node
otherKey: (count:number):string;
}

type LocaleTextProp = Partial<LocaleText>;
```

Comment (dtassone) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:36]: >For i18n instead of 'locale' are you talking about the prop name? Like i18nText or just i18n instead of localeText?

`Locale` or `i18n` we align the prop and the hook to the same name

Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:36]: That should be the case, the hook can do additional logic and can use more that one prop. I see them as 2 separate things, you can have a more generic hook, like `useI18n` that handles the text provided by the user, the locale name etc.

File: packages/grid/_modules_/grid/models/gridOptions.tsx (hunk 17 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/models/gridOptions.tsx:95]: Part of the previous thread - updated

**Review by oliviertassinari — APPROVED** [review]

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 16 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx]: In what way does it handle the locale name?

We could consider 2 hooks...

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx]: I have lost track of the discussion. What problem are we trying to solve?

Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx]: It's not a problem per say but rather - why do we need to match the name of the prop to the name of the hook. Only a few hooks follow this but they are only responsible for one particular feature. I can imagine that for part 2 of that ticket we will have to add additional functionality to the hook and maybe more props if need.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx]: Oh, well, I guess that `localeText` for the prop, `useLocaleText` for the hook, and `getLocaleText` for the getter could make it easier to understand that it's the same "concern".

File: docs/src/pages/components/data-grid/localization/localization.md [hunk 18]
```diff
@@ -7,13 +7,19 @@ components: DataGrid, XGrid
...
-## 🚧 Locale text
+The default locale of Material-UI is English (United States). If you want to use other locales, follow the instructions below.
 
-> ⚠️ This feature isn't implemented yet. It's coming.
->
-> 👍 Upvote [issue #196](https://github.com/mui-org/material-ui-x/issues/196) if you want to see it land faster.
+## Translation keys
+
+You can use the `localeText` prop to pass in your own text and translations.
+You can find all the translation keys supported in [the source](https://github.com/mui-org/material-ui-x/blob/HEAD/packages/grid/_modules_/grid/constants/i18nConstants.ts) in the GitHub repository.
```
Comment (mbrookes) [inline docs/src/pages/components/data-grid/localization/localization.md:15]: We should perhaps document them properly at some point.

File: docs/src/pages/components/data-grid/localization/localization.md [hunk 19]
```diff
@@ -7,13 +7,19 @@ components: DataGrid, XGrid
...
+In the following example, the labels of the density selector are customized.
+
+{{"demo": "pages/components/data-grid/localization/CustomLocaleTextGrid.js", "bg": "inline"}}
+
+## 🚧 Locale text
 
-The grid will support localization with a single prop that includes all the translated messages.
+The data grid will support dozens of locales with a simple import that includes all the translated messages.
 You will be able to follow [this guide](/guides/localization/#locale-text) to use them.
```
Comment (mbrookes) [inline docs/src/pages/components/data-grid/localization/localization.md:23]:
```suggestion
You can follow [this guide](/guides/localization/#locale-text) to use them.
```

File: packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+export const useI18n = (apiRef: ApiRef): void => {
+  const logger = useLogger('useI18n');
+  const { localeText } = useGridSelector(apiRef, optionsSelector);
+
+  const getLocaleText = React.useCallback(
+    (key: any): any => {
+      logger.debug(`Get grid text with key ${key}`);
+
+      if (process.env.NODE_ENV !== 'production' && !localeText.hasOwnProperty(key)) {
+        throw new Error(`Wrong key used for internal purposes ${key}`);
```
Comment (mbrookes) [inline packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx:18]: Sorry, probably me, but what does this mean?

File: packages/grid/_modules_/grid/models/api/i18nApi.ts [hunk 21]
```diff
@@ -0,0 +1,17 @@
+import { LocaleText } from '../gridOptions';
+
+export type LocaleTextValue = string | React.ReactNode | Function;
+
+export type TranslationKeys = keyof LocaleText;
+
+/**
+ * The i18n API interface that is available in the grid [[apiRef]].
```
Comment (mbrookes) [inline packages/grid/_modules_/grid/models/api/i18nApi.ts:8]:
```suggestion
 * The grid i18n API [[apiRef]].
```

What does [[apiRef]] represent?

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 22]
```diff
@@ -87,6 +88,18 @@ export enum DensityTypes {
   Comfortable = 'comfortable',
 }
 
+/**
+ * Sets the types of the texts in the grid.
```
Comment (mbrookes) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:92]:
```suggestion
 * Set the types of the texts in the grid.
```

File: packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx (hunk 20 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/features/i18n/useI18n.tsx:18]: It's for internal usage but I think that we could drop it since TypeScript covers the same use case.

File: packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx (hunk 16 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx]: Ok then, I'll go ahead and rename it 😉

File: docs/src/pages/components/data-grid/localization/localization.md (hunk 18 above)
Comment (DanailH) [reply docs/src/pages/components/data-grid/localization/localization.md:15]: `localeText` prop is added to the api pages of both DataGrid and xGrid. If you are talking about the constant, I've exported an interface with the typing so that people can check it from the code.

Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/localization/localization.md]: > We should perhaps document them properly at some point.

@mbrookes We don't do it either for the core components. I have seen ag-Grid render the default value for the locale. But linking the page on GitHub with the correct version could be as good.

**Review by oliviertassinari — APPROVED** [review]

File: docs/src/pages/components/data-grid/localization/localization.md [hunk 23]
```diff
@@ -7,14 +7,20 @@ components: DataGrid, XGrid
...
-## 🚧 Locale text
+The default locale of Material-UI is English (United States). If you want to use other locales, follow the instructions below.
 
-> ⚠️ This feature isn't implemented yet. It's coming.
->
-> 👍 Upvote [issue #196](https://github.com/mui-org/material-ui-x/issues/196) if you want to see it land faster.
+## Translation keys
+
+You can use the `localeText` prop to pass in your own text and translations.
+You can find all the translation keys supported in [the source](https://github.com/mui-org/material-ui-x/blob/HEAD/packages/grid/_modules_/grid/constants/localeTextConstants.ts) in the GitHub repository.
```
Comment (mbrookes) [inline docs/src/pages/components/data-grid/localization/localization.md:15]: I presume we plan to document this properly at some point?

Comment (DanailH) [reply docs/src/pages/components/data-grid/localization/localization.md:15]: @oliviertassinari answered that in one of the comments before - [https://github.com/mui-org/material-ui-x/pull/718#discussion_r545002613](https://github.com/mui-org/material-ui-x/pull/718#discussion_r545002613)

Maybe we can look into documenting it after all the text is converted, although it may be an overkill since it is also in the code and a reference can work just as good.

File: packages/grid/_modules_/grid/hooks/features/localeText/useLocaleText.tsx [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+  const { localeText } = useGridSelector(apiRef, optionsSelector);
+
+  const getLocaleText = React.useCallback(
+    (key: any): any => {
+      logger.debug(`Get grid text with key ${key}`);
+
+      if (process.env.NODE_ENV !== 'production' && !localeText.hasOwnProperty(key)) {
+        throw new Error(`Wrong key used for internal purposes ${key}`);
+      }
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/localeText/useLocaleText.tsx:20]: TypeScript cover this problem now, no need to duplicate it

```suggestion
```

File: packages/grid/_modules_/grid/constants/localeTextConstants.ts [hunk 25]
```diff
@@ -0,0 +1,7 @@
+export const DEFAULT_LOCALE_TEXT = {
```
Comment (dtassone) [inline packages/grid/_modules_/grid/constants/localeTextConstants.ts:1]: missing type 'LocaleText'

File: packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts [hunk 26]
```diff
@@ -79,6 +80,7 @@ export function useOptionsProp(apiRef: ApiRef, props: GridComponentProps): GridO
       sortModel: props.sortModel,
       density: props.density,
       filterModel: props.filterModel,
+      localeText: { ...DEFAULT_LOCALE_TEXT, ...props.localeText },
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts:83]: that's done using the `DEFAULT_GRID_OPTIONS` merge

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts:83]: in `optionsReducer`

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/utils/useOptionsProp.ts:83]: Unfortunately it doesn't work. It overwrites the existing prop completely and I don't want that, I want a deep merge.
